### DISCUSSION TITLE
location based ar fails after phone goes to sleep and unlocked again

### DISCUSSION THREAD

**Discussion by bildekrokodil** [AR-js-org/AR.js#361] [Help]
Hi all,

I hope somebody can help me out here.

I created a small project. I added a simple box in the street I live in. When I open the website I can see the box in the street.
And walk to it. It is accurate. But.... From the moment I wait to long. The phone is going to stand by. Wich is normal. 
When I unlock the phone again. The website is still running. a few moments later the box jumps way off its position.

I tried it with both. The gps-camera and the gps-projected-camera.

I tried to remove the box with javascript and replace it again so it might simply update its location based on the phones position. But that is not the case. When I refresh the page. It is back to normal.

How can I fix this. I even tried to reset the original coordinates on the camera but nothing works. It also seems like the gps-camera-update-position keeps givving the same lat long and does not update anymore.

Kind regards
Roel

**Comment by nickw1** [comment]
Hello Roel, is the source code available? Obviously if you are prepared to share it, anonymise your test lat and lon. Then I can have a look. This might be related to #278 possibly.

**Comment by bildekrokodil** [comment]
Shure,

This is just the basic example though.

```
<!DOCTYPE html>
<html lang="nl">
    <head>
<meta charset="utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
<title>Test AR</title>
<script src="https://aframe.io/releases/1.2.0/aframe.min.js"></script>
<script src="https://unpkg.com/aframe-look-at-component@0.8.0/dist/aframe-look-at-component.min.js"></script>
<script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script>
    </head>
    <body>

<a-scene
vr-mode-ui="enabled: false"
arjs='sourceType: webcam; videoTexture: true; debugUIEnabled: false;'
renderer='antialias: true; alpha: true' id="ARScene">
<a-camera id="camera01" gps-projected-camera rotation-reader></a-camera>
<a-box material="color: yellow" gps-projected-entity-place="latitude: <lat>; longitude: <long>" position="0 1 0"/>
</a-scene>
    </body>
</html>
```

I might have found a solution to. I dived into the source code of the camera and found a play and pause function.
I added the following just before the </body>. It seems to fix things but I did not tested it very well yet.

````
<script>
    document.onvisibilitychange = function() {
        if (document.visibilityState == "hidden")
            document.getElementById("camera01").components["gps-projected-camera"].pause();
        else
            document.getElementById("camera01").components["gps-projected-camera"].play();
    };
</script>
````

What this does is when the phone is locked. Just before that happens this event is triggered. When the state is hidden. (Phone locked, other tab) I pause the camera. Meaning the gps location watch is removed.
when the page is visible again I call the play function. This will recreate the watch.
Still there is a minor jump in it. But not as big as before. 

It would be nice to recalibrate the origin position back to the phones current location and reposition all entity's based on the long lat they have vs the new origin. Just a idea.

Kind regards
Roel

**Reply by nickw1** [reply]
Ah ok, thanks for that! Let me look at the `watchPosition` behaviour and see if I can fix it, I think I know what the problem might be, it is related to the origin position as you say.

**Comment by nickw1** [comment]
Hi @bildekrokodil I have made some fixes with the way the Web Mercator coordinates are handled. It no longer stores them relative to the original position, but as absolute world coordinates.

 I have also turned off the `maximumAge` setting when watching the position, which has caused problems before. This may help.

Try using https://hikar.org/loc-fix/aframe-ar-nft.js as your AR.js source, and see if that works.

**Reply by nickw1** [reply]
..sorry, you also need `gps-projected-camera` and `gps-projected-entity-place` for the modified code to be used.
